application.py

```diff
@@ -47,50 +59,57 @@ def __init__(self):
         """)
         self.qr_code_label.hide()
         self.layout.addWidget(self.qr_code_label, 0, Qt.AlignBottom)
+
         self.current_video_index = 0
         self.video_list = []
         self.video_data = []
+        self.menu_window = None
+
         self.logs = {
                 "app.load_video_data": "",
                 "app.load_videos": "",
                 "app.play_next_video": "",
                 "app.show_qr_code": "",
             }
+        
         self.video_player.finished.connect(self.fade_out_video) # type: ignore
 
+
     def download_video(self, url, local_video_path: str):
         response = requests.get(url, stream=True)
         if response.status_code == 200:
             logger.info(f"Downloading video {local_video_path}...")
             try:
-                with open(f"{local_video_path}", "wb") as file:
+                with open(f"{local_video_path}.dat.incomplete", "wb") as file:
                     for chunk in response.iter_content(chunk_size=1024*1024):
                         file.write(chunk)
+                os.rename(f"{local_video_path}.dat.incomplete", local_video_path)
                 logger.info(f"Video {local_video_path} downloaded successfully")
             except Exception as e:
                 logger.critical(f"Video could not be downloaded. Error: {e}")
         else:
             logger.critical(f"Video could not be downloaded. Status code: {response.status_code}")
 
-    def load_video_data(self, video_data):
+    def set_video_data(self, video_data):
+        self.video_data = video_data
         try:
-            self.video_data = video_data
-            self.logs["app.load_video_data"] = "Список видео успешно загружен."
+            self.service.save_json(self.video_data, f'{ASSETS_FOLDER}/data/video_data.json')
         except Exception as e:
-            self.logs["app.load_video_data"] = f"Ошибка при загрузке списка видео: {e}"
+            logger.error(e)
 
     def load_videos(self, video_data):
-        for i in self.video_data:
-            local_video_path = os.path.join(f'{ASSETS_FOLDER}/videos/', i['serverfilename'])
-            if not os.path.exists(local_video_path):
-                self.download_video(BASE_URL + "videos/"+ i['serverfilename'], local_video_path)
-            else:
-                logger.info(f"Video {i['serverfilename']} already downloaded.")
-
-
+        self.video_url_list = [BASE_URL + '/videos/' + video['serverfilename'] for video in video_data]
         self.video_list = [video['serverfilename'] for video in video_data]
+
+        for video_url in self.video_url_list:
+            local_video_path = 'assets/videos/' + self.video_list[self.video_url_list.index(video_url)]
```

Comment:
On top of being clunky (which is the main problem), the code you put here works for n^2. Enumerate for the rescue.
```suggestion
        for i, video_url in emumerate(self.video_url_list):
            local_video_path = 'assets/videos/' + self.video_list[i]
```

Reply:
👍